File: src/doogie_perception/matrixhandle.py
#!/usr/bin/env python
# license removed for brevity
import rospy
from doogie_msgs.msg import MazeCellMultiArray
from doogie_msgs.msg import MazeCell
from std_msgs.msg import MultiArrayDimension

class MatrixHandle():
    def __init__(self):
        self.mat = MazeCellMultiArray()
        self.mat.layout.dim.append(MultiArrayDimension())
        self.mat.layout.dim.append(MultiArrayDimension())


    def initMatrix(self, iSize, jSize, iLabel, jLabel, fill = 0, offset = 0):
        self.mat.layout.dim[0].label = iLabel
        self.mat.layout.dim[1].label = jLabel
        self.mat.layout.dim[0].size = iSize
        self.mat.layout.dim[1].size = jSize
        self.mat.layout.dim[0].stride = iSize*jSize
        self.mat.layout.dim[1].stride = jSize
        self.mat.layout.data_offset = offset
        emptyList = []
        for i in range(iSize):
            for j in range(jSize):
                emptyList.append(MazeCell())
        self.mat.data = emptyList


    def getMatrixElement(self, matrix, i, j):
        if type(matrix) is MazeCellMultiArray:
            dstride1 = matrix.layout.dim[1].stride
            offset = matrix.layout.data_offset
            return matrix.data[offset + i + dstride1*j]
        else:
            raise Exception('The array should be an MazeCellMultiArray')
    

    def setMatrixElement(self, matrix, i, j, northElement, southElement, eastElement, westElement, visited):
        if type(matrix) is MazeCellMultiArray:
            dstride1 = matrix.layout.dim[1].stride
            offset = matrix.layout.data_offset
            matrix.data[offset + i + dstride1*j].north_wall = northElement
            matrix.data[offset + i + dstride1*j].south_wall = southElement
            matrix.data[offset + i + dstride1*j].east_wall = eastElement
            matrix.data[offset + i + dstride1*j].west_wall = westElement
            matrix.data[offset + i + dstride1*j].visited = visited

            return matrix.data[offset + i + dstride1*j]
        else:
            raise Exception('The array should be an MazeCellMultiArray')

File: doogie_navigation/src/matrix_handle.cpp
#include <std_msgs/MultiArrayDimension.h>
#include "doogie_navigation/matrix_handle.hpp"

namespace doogie_navigation {

MatrixHandle::MatrixHandle() {
  mat.layout.dim.push_back(std_msgs::MultiArrayDimension());
  mat.layout.dim.push_back(std_msgs::MultiArrayDimension());
}

void MatrixHandle::initMatrix(int i_size, int j_size, std::string i_label, std::string j_label, int fill, int offset) {
  mat.layout.dim[0].size = i_size;
  mat.layout.dim[1].size = j_size;
  mat.layout.dim[0].label = i_label;
  mat.layout.dim[1]. label = j_label;
  mat.layout.dim[0].stride = i_size*j_size;
  mat.layout.dim[1].stride = j_size;
  mat.layout.data_offset = offset;
  doogie_msgs::MazeCellMultiArray empty_list;

  for (std::size_t i = 0; i < i_size; i++) {
      for (std::size_t j = 0; j < j_size; j++) {
          mat.data.push_back(doogie_msgs::MazeCell());
      }
  }
}

doogie_msgs::MazeCell MatrixHandle::getMatrixElement(doogie_msgs::MazeCellMultiArray matrix, int i, int j) {
    int dstride1 = matrix.layout.dim[1].stride;
    int offset = matrix.layout.data_offset;
    return matrix.data[offset + i + dstride1*j];
}

void MatrixHandle::setMatrixElement(doogie_msgs::MazeCellMultiArray matrix, int i, int j, bool north_element,
                                    bool south_element, bool east_element, bool west_element, bool visited) {
  int dstride1 = matrix.layout.dim[1].stride;
  int offset = matrix.layout.data_offset;
  matrix.data[offset + i + dstride1*j].north_wall = north_element;
  matrix.data[offset + i + dstride1*j].south_wall = south_element;
  matrix.data[offset + i + dstride1*j].east_wall = east_element;
  matrix.data[offset + i + dstride1*j].west_wall = west_element;
  matrix.data[offset + i + dstride1*j].visited = visited;
}

}
